src/util/Profiler.h

```diff
@@ -0,0 +1,112 @@
+#pragma once 
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "defines.h"
+
+#define ENABLE_PROFILER
+
+// Helper macros to generate unique* identifier
+// * Limitation: Only one PROFILER_X per line
+// TODO: Use __COUNTER__? It's not standard, but supported on most compilers
+#define _PROFILER_MERGE(A, B) A##B
+#define _PROFILER_LABEL(A) _PROFILER_MERGE(__profiler_, A)
+
+#ifdef ENABLE_PROFILER 
+	#define PROFILE_FUNC() auto _PROFILER_LABEL(__LINE__) = profiler->block(___func___);
+#else 
+	#define PROFILE_FUNC() block;
+#endif
+
+#ifdef ENABLE_PROFILER
+	#define PROFILE_BLOCK(name) auto _PROFILER_LABEL(__LINE__) = profiler->block(name)
+#else 
+	#define PROFILE_BLOCK(name) 
+#endif 
+
+struct ProfileBlock;
+
+class Profiler
+{
+private:
+
+	struct RunStats
+	{
+		uint64_t count;
+		double avg, min, max, last;
+
+		RunStats()
+		{
+			count = 0; avg = 0; min = 99999999.0; max = -1.0; last = 0.0;
+		}
+	};
+
+	struct VectorHasher
+	{
+		std::size_t operator()(const std::vector<std::string>& vec) const
+		{
+			size_t seed = 0;
+			for(const std::string& str : vec)
+			{
+				hash_combine(seed, str);
+			}
+			return seed;
+		}	
+	};
+
+	int compiled;
+	std::unordered_map<std::vector<std::string>, double, VectorHasher> checkpoints;
+	std::unordered_map<std::vector<std::string>, RunStats, VectorHasher> results;
+
+	std::vector<std::string> stack;
```

Comment:
There are several problems with this code.

first of all, you don't want to be storing std::strings. The strings that you are dealing with should be required to be compile-time constants. This allows you to eschew an allocation and copy on every function entry, which is a major performance hit.

Second, if you're going to use std::vector (I do not recommend this), you should use std::stack instead, which provides an actual stack. https://en.cppreference.com/w/cpp/container/stack

Thirdly, you should be implementing this with thread-local storage. https://en.cppreference.com/w/cpp/language/storage_duration

Fourthly, you should probably be using something like std::list, and std::list::splice so that when you pop items off of your stack, the item doesn't get deallocated, but instead gets put into a second std::list that acts as a free-list. This allows you to avoid unnecessary allocations https://en.cppreference.com/w/cpp/container/list/splice

And lastly: there's room here to use some kind of stack allocator (https://github.com/jonesmz/bsf/blob/master/Source/Foundation/bsfUtility/Allocators/BsStackAlloc.h) instead of something like std::list for the things that actually need to be allocated (Hardly anything, I imagine). You're automatically dealing with a stack simply by virtue of how C++ programs work.

So you can have a single thread-local list-head object. This list-head comes into existence at the beginning of your thread. Same for a list-tail.

Each additional scope-item that gets constructed grabs a pointer to the list head and tail, and adds itself to the tail of the list, replacing the pointer for the thread-local list-tail.

When a scope-item's destructor is called, it removes itself from the tail of the list.

If you use doubly-linked-lists, this is trivial to implement.